### ISSUE TITLE
eks sqs reachout, aws credentials question

### ISSUE THREAD

**Issue by michalzuchowski** [zalando-incubator/kube-metrics-adapter#239]
## Expected Behavior
visibility of sqs metrics 

## Actual Behavior
kubectl describe hpa custom-metrics-consumer:

unable to get external metric default/app-queue-length/&LabelSelector{MatchLabels:map[string]string{queue-name: seg_query_eng-query_exec-us-east-1-michalzuchowski-req.fifo,region: us-east-1,type: sqs-queue-length,},MatchExpressions:[]LabelSelectorRequirement{},}: no metrics returned from external metrics API

## Question
In https://github.com/zalando-incubator/kube-metrics-adapter/blob/master/README.md#aws-iam-role
what exactly mean: "To integrate with AWS, the controller needs to run on nodes with access to AWS API."
Is this mean: 
- node can reach out to api through network/firewall (that was my first tought)
or
- you have to have actual aws credentials for api like here: https://github.com/zalando-incubator/kubernetes-on-aws/blob/6d7835a5906847837ec905e43573eb8b9e8dc0f6/cluster/manifests/kube-metrics-adapter/deployment.yaml#L33-L35

**Comment by michalzuchowski** [comment]
I've just created aws credentials from secrert, and I'm able to get metrics from kube-metrics-adapter:

 kubectl exec --stdin --tty kube-metrics-adapter-86ccd64f7-8schz -n kube-system -- /bin/sh 
/ # aws --region us-east-1 cloudwatch get-metric-statistics --metric-name ApproximateNumberOfMessagesVisible --namespace AWS/SQS --period 3600 --start-time 2020-12-22T11:00:00Z --end-time 2020-12-22T13:00:00Z --dimen
sions Name=QueueName,Value=seg_query_eng-query_exec-us-east-1-michalzuchowski-req.fifo --statistics Average
{
    "Label": "ApproximateNumberOfMessagesVisible",
    "Datapoints": [
        {
            "Timestamp": "2020-12-22T11:00:00Z",
            "Average": 3.0,
            "Unit": "Count"
        },
        {
            "Timestamp": "2020-12-22T12:00:00Z",
            "Average": 3.0,
            "Unit": "Count"
        }
    ]
}
/ # 
But  kubectl describe hpa custom-metrics-consumer:
ScalingActive  False   FailedGetExternalMetric  the HPA was unable to compute the replica count: unable to get external metric default/app-queue-length/&LabelSelector{MatchLabels:map[string]string{queue-name: seg_query_eng-query_exec-us-east-1-michalzuchowski-req.fifo,region: us-east-1,type: sqs-queue-length,},MatchExpressions:[]LabelSelectorRequirement{},}: no metrics returned from external metrics API

Here is hpa:

kubectl get hpa custom-metrics-consumer -o json
{
    "apiVersion": "autoscaling/v1",
    "kind": "HorizontalPodAutoscaler",
    "metadata": {
        "annotations": {
            "autoscaling.alpha.kubernetes.io/conditions": "[{\"type\":\"AbleToScale\",\"status\":\"True\",\"lastTransitionTime\":\"2020-12-29T15:01:36Z\",\"reason\":\"SucceededGetScale\",\"message\":\"the HPA controller was able to get the target's current scale\"},{\"type\":\"ScalingActive\",\"status\":\"False\",\"lastTransitionTime\":\"2020-12-29T15:01:36Z\",\"reason\":\"FailedGetExternalMetric\",\"message\":\"the HPA was unable to compute the replica count: unable to get external metric default/app-queue-length/\\u0026LabelSelector{MatchLabels:map[string]string{queue-name: seg_query_eng-query_exec-us-east-1-michalzuchowski-req.fifo,region: us-east-1,type: sqs-queue-length,},MatchExpressions:[]LabelSelectorRequirement{},}: no metrics returned from external metrics API\"}]",
            "autoscaling.alpha.kubernetes.io/metrics": "[{\"type\":\"External\",\"external\":{\"metricName\":\"app-queue-length\",\"metricSelector\":{\"matchLabels\":{\"queue-name\":\"seg_query_eng-query_exec-us-east-1-michalzuchowski-req.fifo\",\"region\":\"us-east-1\",\"type\":\"sqs-queue-length\"}},\"targetAverageValue\":\"2\"}}]",
            "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"autoscaling/v2beta2\",\"kind\":\"HorizontalPodAutoscaler\",\"metadata\":{\"annotations\":{},\"labels\":{\"application\":\"custom-metrics-consumer\"},\"name\":\"custom-metrics-consumer\",\"namespace\":\"default\"},\"spec\":{\"maxReplicas\":10,\"metrics\":[{\"external\":{\"metric\":{\"name\":\"app-queue-length\",\"selector\":{\"matchLabels\":{\"queue-name\":\"seg_query_eng-query_exec-us-east-1-michalzuchowski-req.fifo\",\"region\":\"us-east-1\",\"type\":\"sqs-queue-length\"}}},\"target\":{\"averageValue\":\"2\",\"type\":\"AverageValue\"}},\"type\":\"External\"}],\"minReplicas\":1,\"scaleTargetRef\":{\"apiVersion\":\"apps/v1\",\"kind\":\"Deployment\",\"name\":\"custom-metrics-consumer\"}}}\n"
        },
        "creationTimestamp": "2020-12-29T15:01:21Z",
        "labels": {
            "application": "custom-metrics-consumer"
        },
        "managedFields": [
            {
                "apiVersion": "autoscaling/v2beta2",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:metadata": {
                        "f:annotations": {
                            ".": {},
                            "f:kubectl.kubernetes.io/last-applied-configuration": {}
                        },
                        "f:labels": {
                            ".": {},
                            "f:application": {}
                        }
                    },
                    "f:spec": {
                        "f:maxReplicas": {},
                        "f:metrics": {},
                        "f:minReplicas": {},
                        "f:scaleTargetRef": {
                            "f:apiVersion": {},
                            "f:kind": {},
                            "f:name": {}
                        }
                    }
                },
                "manager": "kubectl",
                "operation": "Update",
                "time": "2020-12-29T15:01:21Z"
            },
            {
                "apiVersion": "autoscaling/v1",
                "fieldsType": "FieldsV1",
                "fieldsV1": {
                    "f:metadata": {
                        "f:annotations": {
                            "f:autoscaling.alpha.kubernetes.io/conditions": {}
                        }
                    },
                    "f:status": {
                        "f:currentReplicas": {}
                    }
                },
                "manager": "kube-controller-manager",
                "operation": "Update",
                "time": "2020-12-29T15:01:37Z"
            }
        ],
        "name": "custom-metrics-consumer",
        "namespace": "default",
        "resourceVersion": "6213128",
        "selfLink": "/apis/autoscaling/v1/namespaces/default/horizontalpodautoscalers/custom-metrics-consumer",
        "uid": "e91d498c-2854-4300-8194-d823fe4f1d57"
    },
    "spec": {
        "maxReplicas": 10,
        "minReplicas": 1,
        "scaleTargetRef": {
            "apiVersion": "apps/v1",
            "kind": "Deployment",
            "name": "custom-metrics-consumer"
        }
    },
    "status": {
        "currentReplicas": 10,
        "desiredReplicas": 0
    }
}

**Comment by mikkeloscar** [comment]
Is there anything in the logs of the `kube-metrics-adapter` maybe it suggests what is going on?

**Comment by michalzuchowski** [comment]
@mikkeloscar 
Everything looks as should be, except that kube-metrics-adapter can't get any sqs metrics.
Crucial question is:

In https://github.com/zalando-incubator/kube-metrics-adapter/blob/master/README.md#aws-iam-role
what exactly mean: "To integrate with AWS, the controller needs to run on nodes with access to AWS API."
Is this mean:

node can reach out to api through network/firewall (that was my first tought)
or
you have to have actual aws credentials for api like here: https://github.com/zalando-incubator/kubernetes-on-aws/blob/6d7835a5906847837ec905e43573eb8b9e8dc0f6/cluster/manifests/kube-metrics-adapter/deployment.yaml#L33-L35

Logs from kube-metrics-adapter pod in kube-system namespace:

`time="2021-01-04T09:48:49Z" level=info msg="Looking for HPAs" provider=hpa
time="2021-01-04T09:48:49Z" level=info msg="Removing previously scheduled metrics collector: {custom-metrics-consumer default}" provider=hpa
time="2021-01-04T09:48:49Z" level=info msg="Found 1 new/updated HPA(s)" provider=hpa
time="2021-01-04T09:48:49Z" level=info msg="Event(v1.ObjectReference{Kind:\"HorizontalPodAutoscaler\", Namespace:\"default\", Name:\"custom-metrics-consumer\", UID:\"a9e90e58-e0a2-4aac-bc71-30aec8370146\", APIVersion:\"autoscaling/v2beta2\", ResourceVersion:\"6900520\", FieldPath:\"\"}): type: 'Warning' reason: 'CreateNewMetricsCollector' Failed to create new metrics collector: metric 'app-queue-length' not supported"
`
kube-metrics-adapter deploymeny description

`kubectl describe deployment kube-metrics-adapter -n kube-system
Name:                   kube-metrics-adapter
Namespace:              kube-system
CreationTimestamp:      Thu, 31 Dec 2020 12:10:50 +0100
Labels:                 application=kube-metrics-adapter
                        version=latest
Annotations:            deployment.kubernetes.io/revision: 1
                        kubectl.kubernetes.io/last-applied-configuration:
                          {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"application":"kube-metrics-adapter","version":"latest"...
Selector:               application=kube-metrics-adapter
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           application=kube-metrics-adapter
                    version=latest
  Annotations:      iam.amazonaws.com/role: seg_query_eng-query_exec-us-east-1-michalzuchowski-eks
  Service Account:  custom-metrics-apiserver
  Containers:
   kube-metrics-adapter:
    Image:      registry.opensource.zalan.do/teapot/kube-metrics-adapter:latest
    Port:       <none>
    Host Port:  <none>
    Args:
      --aws-external-metrics
    Limits:
      cpu:     100m
      memory:  100Mi
    Requests:
      cpu:     100m
      memory:  100Mi
    Environment:
      AWS_REGION:                   us-east-1
      AWS_SHARED_CREDENTIALS_FILE:  /meta/aws-iam/credentials
      AWS_PROFILE:                  DevelopmentCampaign_Administrator
    Mounts:
      /meta/aws-iam from aws-iam-credentials (ro)
  Volumes:
   aws-iam-credentials:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kube-metrics-adapter-aws-iam-credentials
    Optional:    false
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
  Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  <none>
NewReplicaSet:   kube-metrics-adapter-685f4f95cb (1/1 replicas created)
Events:          <none>

`

**Comment by mikkeloscar** [comment]
> what exactly mean: "To integrate with AWS, the controller needs to run on nodes with access to AWS API."
Is this mean:

Sorry, this is a bit confusing the way it's written. Basically we mean that kube-metrics-adapter needs to have the right AWS credentials for sqs (however you want to ensure that).


In the logs we see:

```
time="2021-01-04T09:48:49Z" level=info msg="Event(v1.ObjectReference{Kind:\"HorizontalPodAutoscaler\", Namespace:\"default\", Name:\"custom-metrics-consumer\", UID:\"a9e90e58-e0a2-4aac-bc71-30aec8370146\", APIVersion:\"autoscaling/v2beta2\", ResourceVersion:\"6900520\", FieldPath:\"\"}): type: 'Warning' reason: 'CreateNewMetricsCollector' Failed to create new metrics collector: metric 'app-queue-length' not supported"
```

Which suggests the HPA metric is not being correctly configured. You need to also specify the flag `--aws-region=<region>` otherwise the sqs collector won't be enabled. Similar to how we run it here (change the region to the one you use for sqs): https://github.com/zalando-incubator/kubernetes-on-aws/blob/58c801754aea3ee22e1eeef35cfc915c635306c5/cluster/manifests/kube-metrics-adapter/deployment.yaml#L46

**Comment by michalzuchowski** [comment]
@mikkeloscar 
I've added ```--aws-region to ``` parameters: kube-metrics-adapter

Here are my hpa code and logs.
 This reduced to sqs is copy from https://github.com/zalando-incubator/kube-metrics-adapter/blob/master/example/deploy/hpa.yaml

Can you glance

```
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  name: custom-metrics-consumer
  namespace: default
  labels:
    application: custom-metrics-consumer
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: custom-metrics-consumer
  minReplicas: 1
  maxReplicas: 10
  metrics:

  - type: External
    external:
      metric:
        name: app-queue-length
        selector:
          matchLabels:
            type: sqs-queue-length
            queue-name: seg_query_eng-query_exec-us-east-1-michalzuchowski-req.fifo
            region: us-east-1
      target:
        averageValue: "2"
        type: AverageValue
```

hpa logs:

```
Conditions:
  Type           Status  Reason                   Message
  ----           ------  ------                   -------
  AbleToScale    True    SucceededGetScale        the HPA controller was able to get the target's current scale
  ScalingActive  False   FailedGetExternalMetric  the HPA was unable to compute the replica count: unable to get external metric default/app-queue-length/&LabelSelector{MatchLabels:map[string]string{queue-name: seg_query_eng-query_exec-us-east-1-michalzuchowski-req.fifo,region: us-east-1,type: sqs-queue-length,},MatchExpressions:[]LabelSelectorRequirement{},}: no metrics returned from external metrics API
Events:
  Type     Reason                        Age   From                       Message
  ----     ------                        ----  ----                       -------
  Warning  FailedGetExternalMetric       2s    horizontal-pod-autoscaler  unable to get external metric default/app-queue-length/&LabelSelector{MatchLabels:map[string]string{queue-name: seg_query_eng-query_exec-us-east-1-michalzuchowski-req.fifo,region: us-east-1,type: sqs-queue-length,},MatchExpressions:[]LabelSelectorRequirement{},}: no metrics returned from external metrics API
  Warning  FailedComputeMetricsReplicas  2s    horizontal-pod-autoscaler  invalid metrics (1 invalid out of 1), first error is: failed to get app-queue-length external metric: unable to get external metric default/app-queue-length/&LabelSelector{MatchLabels:map[string]string{queue-name: seg_query_eng-query_exec-us-east-1-michalzuchowski-req.fifo,region: us-east-1,type: sqs-queue-length,},MatchExpressions:[]LabelSelectorRequirement{},}: no metrics returned from external metrics API
```

**Comment by mikkeloscar** [comment]
Thanks for all the information. Based on this I found a bug fixed in #251

Can you try this docker image: `registry.opensource.zalan.do/teapot/kube-metrics-adapter-test:pr-251-1` (built from the PR) and validate that the issue is fixed for you with that?

**Comment by michalzuchowski** [comment]
Not helped, but I'm going to check this on differnt eks cluster
